# SPDX-FileCopyrightText: 2023 Greenbone AG
# Some text descriptions might be excerpted from (a) referenced
# source(s), and are Copyright (C) by the respective right holder(s).
#
# SPDX-License-Identifier: GPL-2.0-only

CPE = "cpe:/a:python:python";

if(description)
{
  script_oid("1.3.6.1.4.1.25623.1.0.170753");
  script_version("2023-12-26T05:05:23+0000");
  script_tag(name:"last_modification", value:"2023-12-26 05:05:23 +0000 (Tue, 26 Dec 2023)");
  script_tag(name:"creation_date", value:"2023-12-22 10:41:08 +0000 (Fri, 22 Dec 2023)");
  script_tag(name:"cvss_base", value:"6.1");
  script_tag(name:"cvss_base_vector", value:"AV:N/AC:L/Au:M/C:N/I:C/A:N");
  script_tag(name:"severity_vector", value:"CVSS:3.1/AV:N/AC:L/PR:H/UI:N/S:U/C:N/I:H/A:N");
  script_tag(name:"severity_origin", value:"NVD");
  script_tag(name:"severity_date", value:"2023-12-13 20:17:00 +0000 (Wed, 13 Dec 2023)");

  script_cve_id("CVE-2023-6507");

  script_tag(name:"qod_type", value:"remote_banner_unreliable");

  script_tag(name:"solution_type", value:"VendorFix");

  script_name("Python 3.12.0 Improper Privilege Management Vulnerability - Linux");

  script_category(ACT_GATHER_INFO);

  script_copyright("Copyright (C) 2023 Greenbone AG");
  script_family("General");
  script_dependencies("gb_python_consolidation.nasl", "os_detection.nasl");
  script_mandatory_keys("python/detected", "Host/runs_unixoide");

  script_tag(name:"summary", value:"CPython is prone to an improper privilege management
  vulnerability.");

  script_tag(name:"vuldetect", value:"Checks if a vulnerable version is present on the target host.");

  script_tag(name:"insight", value:"When using the extra_groups= parameter with an empty list as a
  value (ie extra_groups=[]) the logic regressed to not call setgroups(0, NULL) before calling exec(),
  thus not dropping the original processes' groups before starting the new process. There is no issue
  when the parameter isn't used or when any value is used besides an empty list.

  This issue only impacts CPython processes run with sufficient privilege to make the setgroups
  system call (typically root).");

  script_tag(name:"affected", value:"CPython version 3.12.0.");

  script_tag(name:"solution", value:"Update to version 3.12.1 or later.");

  script_xref(name:"URL", value:"https://mail.python.org/archives/list/security-announce@python.org/thread/AUL7QFHBLILGISS7U63B47AYSSGJJQZD/");
  script_xref(name:"URL", value:"https://github.com/python/cpython/issues/112334");

  exit(0);
}

include("host_details.inc");
include("version_func.inc");

if( isnull( port = get_app_port( cpe:CPE ) ) )
  exit( 0 );

if( ! infos = get_app_version_and_location( cpe:CPE, port:port, exit_no_version:TRUE, version_regex:"^[0-9]+\.[0-9]+\.[0-9]+" ) )
  exit( 0 );

version = infos["version"];
location = infos["location"];

if ( version_is_equal( version: version, test_version: "3.12.0" ) ) {
  report = report_fixed_ver( installed_version: version, fixed_version: "3.12.1", install_path: location );
  security_message( port: port, data: report );
  exit( 0 );
}

exit( 99 );
